pkg/scheduler/predicates/ha.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package predicates
+
+import (
+	"fmt"
+
+	"github.com/pingcap/tidb-operator/pkg/label"
+	"github.com/pingcap/tidb-operator/pkg/util"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	replicas int32 = 3
+)
+
+type ha struct {
+	kubeCli   kubernetes.Interface
+	podListFn func(ns, clusterName, component string) (*apiv1.PodList, error)
+}
+
+// NewHA returns a Predicate
+func NewHA(kubeCli kubernetes.Interface) Predicate {
+	h := &ha{
+		kubeCli: kubeCli,
+	}
+	h.podListFn = h.realPodListFn
+	return h
+}
+
+func (h *ha) Name() string {
+	return "HighAvailability"
+}
+
+func (h *ha) Filter(clusterName string, pod *apiv1.Pod, nodes []apiv1.Node) ([]apiv1.Node, error) {
```

Comment:
Add a short description of the algrithom how we schedule pods.

1. First, we sort all the nodes we get from kube-scheduler by how many same kind of pod it contains, find the nodes that have least pods.
2. When scheduling the first `replicas` pods, we must ensure no previous pods on the nodes.
3. For later pods, we choose the nodes that have least pods.